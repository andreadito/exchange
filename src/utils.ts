export enum Currency {
    EUR = 'EUR',
    GBP = 'GBP',
    USD = 'USD',
}

export interface WalletData {
    balance: number;
    currency: Currency;
}

export interface Wallet {
    id: string,
    data: WalletData
}

export interface Rate {
    symbol: string;
    value: string;
}

export const MIN_ALLOWED = 0.5;

export function convertValue(value: number, rate: number): number {
    return value * rate;
}

export function limitDecimal(value: string) {
    return (value.indexOf(".") >= 0) ? (value.substr(0, value.indexOf(".")) + value.substr(value.indexOf("."), 3)) : value
}
