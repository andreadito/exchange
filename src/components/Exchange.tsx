import {Box, Button, Form, Text} from "grommet";
import {PowerCycle} from "grommet-icons";
import React, {Fragment, useEffect, useReducer} from "react";
import ExchangeInput from "./ExchangeInput";
import {useForm} from "react-hook-form";
import { useRate } from "../hooks/useRate";
import {
    exchangeReducer,
    SET_BASE, SET_EXCHANGE, SET_IS_EXCEEDING,
    SET_RATE_FROM,
    SET_RATE_TO,
    SET_SYMBOL
} from "../reducers/exchange";
import {convertValue, MIN_ALLOWED, Wallet} from "../utils";
import Rates from "./Rates";
import {useExchange} from "../hooks/useExchange";

export interface ExchangeProps {
    wallets: Wallet[];
}

type Inputs = {
    from_wallet: Wallet,
    from_value: number,
    to_wallet: Wallet,
    to_value: number
};

export default function Exchange({ wallets }: ExchangeProps) {

    const initialState = {
        base: wallets[0].data.currency,
        symbol: wallets[1].data.currency,
        rateFrom: 1,
        rateTo: 1,
        isExceeding: false,
        exchange: {}
    }

    const [state, dispatch] = useReducer(exchangeReducer, initialState);

    const initialFromWallet = wallets[0];
    const initialToWallet = wallets[1];

    const defaultValues = {
        from_value: 0,
        from_wallet: initialFromWallet,
        to_value: 0,
        to_wallet: initialToWallet,
    };

    const { handleSubmit, control, errors, setValue, getValues, watch } = useForm<Inputs>({
        defaultValues,
        mode: "all"
    });

    const { isLoading: rateInfoLoading, data: rateInfo } = useRate(state.base, state.symbol);

    const [mutate, {isLoading: isExchageIsLoading, isSuccess: exchangeSuccess}] = useExchange();

    const { from_wallet, from_value, to_wallet, to_value } = watch();

    const onSubmitForm = async (data: Inputs) => {
        try {
            await mutate({operation: 'sub', value: from_value, walletid: from_wallet.id});
            await mutate({operation: 'add', value: parseFloat(to_value.toFixed(2)), walletid: to_wallet.id});
            dispatch({type: SET_EXCHANGE, payload: {
                    from_value, from_currency:
                    from_wallet.data.currency,
                    to_value: parseFloat(to_value.toFixed(2)),
                    to_currency: to_wallet.data.currency
            }})
        } catch (error) {
            console.log(error);
        }
    }


    const onErrorForm = (errors: any) => console.log(errors);

    useEffect(() => {
        if(from_value > MIN_ALLOWED) {
            setValue('to_value', convertValue(from_value, state.rateTo))
        } else {
            setValue('to_value', 0)
        }

        dispatch({type: SET_BASE, payload: from_wallet.data.currency})
        dispatch({type: SET_SYMBOL, payload: to_wallet.data.currency})

        if (rateInfo && rateInfo.data.rates[to_wallet.data.currency]) {
            dispatch({type: SET_RATE_TO, payload: rateInfo.data.rates[to_wallet.data.currency]})
            dispatch({type: SET_RATE_FROM, payload: rateInfo.data.rates[from_wallet.data.currency] })
        }

        dispatch({type: SET_IS_EXCEEDING, payload: from_wallet.data.balance < from_value })

    }, [from_value, from_wallet.data.balance, from_wallet.data.currency, rateInfo, setValue, state.rateTo, to_value, to_wallet.data.currency])

    const swapFields = () => {
        const {from_wallet, to_wallet} = getValues();

        setValue('from_wallet', to_wallet);
        setValue('to_wallet', from_wallet)
    }

    const disableSubmit = Object.values(errors).length > 0
        || state.isExceeding
        || (from_value && from_value < MIN_ALLOWED)
        || state.base === state.symbol
        || isExchageIsLoading

    return (
        <Box animation={"fadeIn"}>
            {!exchangeSuccess ? <Form onSubmit={handleSubmit(onSubmitForm, onErrorForm)}>
                <Box gap={'large'}>
                    <ExchangeInput
                        name={'from'}
                        label={'From'}
                        wallets={wallets}
                        control={control}
                        errors={errors}
                        rate={state.rateFrom}
                        isExceeding={state.isExceeding}
                    />
                    <ExchangeInput
                        name={'to'}
                        label={'To'}
                        wallets={wallets}
                        control={control}
                        errors={errors}
                        rate={state.rateTo}
                        readonly={true}
                    />

                    <Rates symbol={state.symbol} base={state.base} rateTo={state.rateTo} isLoading={rateInfoLoading}/>

                    <Box direction={'row'} gap={'medium'} justify={"center"}>
                        <Button onClick={swapFields}>
                            <PowerCycle size={'medium'} />
                        </Button>
                        <Button disabled={disableSubmit} size={'large'} primary label="Exchange" type={'submit'} />
                    </Box>

                </Box>
            </Form> : <Fragment>
                <Text size={'large'}>You've exchanged: {state.exchange.from_value} {state.exchange.from_currency} in {state.exchange.to_value} {state.exchange.to_currency}</Text>
            </Fragment>}
        </Box>
    )
}
