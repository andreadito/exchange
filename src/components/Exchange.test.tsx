import React from "react";
import {Currency, Wallet} from "../utils";
import Exchange from "./Exchange";
import renderer from 'react-test-renderer';
import 'mutationobserver-shim';

describe("Exchange", () => {
    const wallets: Wallet[] = [
        {
            "id": "AyMzv3EkKFzFG4H8HfEz",
            "data": {
                "currency": Currency.GBP,
                "balance": 1090
            }
        },
        {
            "id": "SPNFH5Q5pDLBbYOPrOd6",
            "data": {
                "currency": Currency.EUR,
                "balance": 0
            }
        },
        {
            "id": "amAienpx5MVa5EJkyJEc",
            "data": {
                "balance": 10,
                "currency": Currency.USD
            }
        }
    ];

    it("true", () => {
        const tree = renderer.create(<Exchange wallets={wallets}/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
})
