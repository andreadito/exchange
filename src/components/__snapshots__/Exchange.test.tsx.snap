// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange true 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 hnasWA"
>
  <form
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="StyledBox-sc-13pk1d4-0 gAsOLC"
    >
      <div
        className="StyledBox-sc-13pk1d4-0 gAsOLC"
      >
        <div
          className="StyledBox-sc-13pk1d4-0 fkEaZX"
        >
          <div
            className="StyledBox-sc-13pk1d4-0 kJuWxj"
          >
            <button
              aria-label="Open Drop"
              className="StyledButton-sc-323bzc-0 kUDdtj Select__StyledSelectDropButton-sc-17idtfo-1 bsheiM"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              type="button"
            >
              <div
                className="StyledBox-sc-13pk1d4-0 lhHpuy"
              >
                <div
                  className="StyledBox-sc-13pk1d4-0 gErjJY"
                >
                  <div
                    className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
                  >
                    <input
                      autoComplete="off"
                      className="StyledTextInput-sc-1x30a0s-0 eymfXE Select__SelectTextInput-sc-17idtfo-0 jwiVPE"
                      name="from_wallet"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="From"
                      readOnly={true}
                      size="medium"
                      tabIndex="-1"
                      type="text"
                      value="GBP"
                    />
                  </div>
                </div>
                <div
                  className="StyledBox-sc-13pk1d4-0 dderd"
                  style={
                    Object {
                      "minWidth": "auto",
                    }
                  }
                >
                  <svg
                    aria-label="FormDown"
                    className="StyledIcon-ofa7kd-0 gvGxfc"
                    viewBox="0 0 24 24"
                  >
                    <polyline
                      fill="none"
                      points="18 9 12 15 6 9"
                      stroke="#000"
                      strokeWidth="2"
                    />
                  </svg>
                </div>
              </div>
            </button>
            <span
              className="StyledText-sc-1sadyjn-0 CVDRa"
              style={
                Object {
                  "marginTop": 10,
                }
              }
            >
              Balance: 1090.00
            </span>
          </div>
          <div
            className="StyledBox__StyledBoxGap-sc-13pk1d4-1 bZqeeV"
          />
          <div
            className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
          >
            <input
              autoComplete="off"
              className="StyledTextInput-sc-1x30a0s-0 ksHnCo"
              inputMode="decimal"
              name="from_value"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              placeholder="0"
              size="medium"
              step={0.01}
              style={
                Object {
                  "textAlign": "right",
                }
              }
              type="number"
              value={0}
            />
          </div>
        </div>
      </div>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 qCliw"
      />
      <div
        className="StyledBox-sc-13pk1d4-0 gAsOLC"
      >
        <div
          className="StyledBox-sc-13pk1d4-0 fkEaZX"
        >
          <div
            className="StyledBox-sc-13pk1d4-0 kJuWxj"
          >
            <button
              aria-label="Open Drop"
              className="StyledButton-sc-323bzc-0 kUDdtj Select__StyledSelectDropButton-sc-17idtfo-1 eoEKMM"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              type="button"
            >
              <div
                className="StyledBox-sc-13pk1d4-0 lhHpuy"
              >
                <div
                  className="StyledBox-sc-13pk1d4-0 gErjJY"
                >
                  <div
                    className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
                  >
                    <input
                      autoComplete="off"
                      className="StyledTextInput-sc-1x30a0s-0 eymfXE Select__SelectTextInput-sc-17idtfo-0 jwiVPE"
                      name="to_wallet"
                      onBlur={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="To"
                      readOnly={true}
                      size="medium"
                      tabIndex="-1"
                      type="text"
                      value="EUR"
                    />
                  </div>
                </div>
                <div
                  className="StyledBox-sc-13pk1d4-0 dderd"
                  style={
                    Object {
                      "minWidth": "auto",
                    }
                  }
                >
                  <svg
                    aria-label="FormDown"
                    className="StyledIcon-ofa7kd-0 gvGxfc"
                    viewBox="0 0 24 24"
                  >
                    <polyline
                      fill="none"
                      points="18 9 12 15 6 9"
                      stroke="#000"
                      strokeWidth="2"
                    />
                  </svg>
                </div>
              </div>
            </button>
            <span
              className="StyledText-sc-1sadyjn-0 CVDRa"
              style={
                Object {
                  "marginTop": 10,
                }
              }
            >
              Balance: 0.00
            </span>
          </div>
          <div
            className="StyledBox__StyledBoxGap-sc-13pk1d4-1 bZqeeV"
          />
          <div
            className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
          >
            <input
              autoComplete="off"
              className="StyledTextInput-sc-1x30a0s-0 eymfXE"
              inputMode="decimal"
              name="to_value"
              onBlur={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              placeholder="0"
              readOnly={true}
              size="medium"
              step={0.01}
              style={
                Object {
                  "textAlign": "right",
                }
              }
              type="number"
              value="0.00"
            />
          </div>
        </div>
      </div>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 qCliw"
      />
      <div
        className="StyledBox-sc-13pk1d4-0 iIIbdL"
        style={
          Object {
            "borderColor": "#6FFFB0",
            "borderRadius": "10px",
            "borderStyle": "solid",
            "borderWidth": "small",
          }
        }
      >
        <span
          className="StyledText-sc-1sadyjn-0 jhkZDB"
          role="rate_value"
        >
          Loading Rates...
        </span>
      </div>
      <div
        className="StyledBox__StyledBoxGap-sc-13pk1d4-1 qCliw"
      />
      <div
        className="StyledBox-sc-13pk1d4-0 hqpbFa"
      >
        <button
          className="StyledButton-sc-323bzc-0 kUDdtj"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="button"
        >
          <svg
            aria-label="PowerCycle"
            className="StyledIcon-ofa7kd-0 bzzphA"
            viewBox="0 0 24 24"
          >
            <path
              d="M20,8 C18.5343681,5.03213345 15.4860999,3 11.9637942,3 C7.01333514,3 3,7.02954545 3,12 M4,16 C5.4656319,18.9678666 8.51390007,21 12.0362058,21 C16.9866649,21 21,16.9704545 21,12 M9,16 L3,16 L3,22 M21,2 L21,8 L15,8"
              fill="none"
              stroke="#000"
              strokeWidth="2"
            />
          </svg>
        </button>
        <div
          className="StyledBox__StyledBoxGap-sc-13pk1d4-1 eHjkEI"
        />
        <button
          className="StyledButton-sc-323bzc-0 bYHuRP"
          disabled={false}
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          type="submit"
        >
          Exchange
        </button>
      </div>
    </div>
  </form>
</div>
`;
